jobserver/settings.py

```diff
@@ -34,6 +34,21 @@
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = env.str("SECRET_KEY")
 
+
+def configure_secret_key_fallbacks():
+    """Return a list containing a single secret key for SECRET_KEY_FALLBACKS."""
+    old_secret_key = env.str("OLD_SECRET_KEY", default=None)
+    if old_secret_key is not None:
+        return [old_secret_key]
+
+    # Empty list is the default Django value.
+    return []
```

Comment:
I'm not completely sure if the extra complexity is worth the test here.

The simplest way of doing this may be sufficient:

```diff
diff --git a/jobserver/settings.py b/jobserver/settings.py
index b4fad626..96e35836 100644
--- a/jobserver/settings.py
+++ b/jobserver/settings.py
@@ -34,6 +34,12 @@ BASE_DIR = Path(__file__).resolve().parent.parent
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = env.str("SECRET_KEY")
 
+# Optional fallback for rotating the secret key.
+# Any OLD_SECRET_KEY that is used should eventually be removed.
+OLD_SECRET_KEY = env.str("OLD_SECRET_KEY", default=None)
+if OLD_SECRET_KEY is not None:
+    SECRET_KEY_FALLBACKS = [OLD_SECRET_KEY]
+
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = env.bool("DEBUG", default=False)
```

That said, with it being a security configuration, it may make sense to test it behaves as expected.

The reason for the function is to make the change easier to test: otherwise, you'd have to import the whole `settings.py`. Even if you import the module `importlib` to avoid reloading the existing `settings.py` in the global namespace, there are still other side-effects, because `settings.py` doesn't only set module-level variables. It also runs `initialise_sentry()`.

Reply:
I agree.

This seems to be in a function to help with testing, but our tests don't actually test the behaviour we care about, which is that a user who's previously logged in when the key was X does not have to log in now that the key is Y.

Reply:
An excellent point; not one I had considered 🙂

Reply:
> This seems to be in a function to help with testing, but our tests don't actually test the behaviour we care about, which is that a user who's previously logged in when the key was X does not have to log in now that the key is Y.

We'd then be testing Django's behaviour even more so, so the simple option is preferable.